### PR TITLE
Add support for typeDateTim4 and typeDateTime in switch statements

### PR DESCRIPTION
**Pull request by forfuncsake**
I found two more column data types that weren't fully supported in the makeGoLangScanType func and associated switch statements, which result in panics.

SMALLDATETIME NOT NULL and DATETIME NOT NULL

### CODE REVIEW

**Comment by forfuncsake** [comment]
I just noticed a partially conflicting PR in #302, so imported those related changes into this branch.

I also used the string type "BINARY" for typeBigBinary instead of "TIMESTAMP", to comply with https://msdn.microsoft.com/en-us/library/dd358341.aspx. Does that still work for your timestamp use case @brown-m?

**Comment by brown-m** [comment]
@forfuncsake i'm in two minds with respect to the MSDN spec. to my mind, a binary data type is a general storage case with the responsibility to track how to encode / decode the stored data falling on the application developer, for instance, storing binary data for PNGs (a crap idea not recommended). whereas TIMESTAMP, whilst technically binary data, an application designer has no input as to what data gets stored in it as it's populated by the DB. with that in mind i would expect the library to handle it as a TIMESTAMP instead of a generalised binary type.

simply put, and I may be in the minority with this wish, i would expect the use of ColumnTypes() to give me comparable results to running a ```select``` statement on INFORMATION_SCHEMA which does return TIMESTAMP for the column type instead of a generic BINARY field. but one can dream...

i believe the intent and usage for the ColumnTypes() function should be clarified before deciding on a definitive course of action.

**Comment by forfuncsake** [comment]
and/or @kardianos

**Comment by kardianos** [comment]
You opened this a while ago. I'm going to try to get through the backlog. If you have time, can you add a test that would fail before and does not fail with this PR?

**Comment by forfuncsake** [comment]
Done.
